### PR TITLE
feat(ops-hardening): surface and repair dead summary-job backlog 

### LINKED ISSUE #901: Dead summary-job backlog is hidden from normal status/health and lacks a safe repair path
## Summary

A live `0.147.1` database contained:

```text
completed summary jobs: 92
dead summary jobs:      1,667
```

Normal status and health surfaces did not make this backlog obvious.

## Expected behavior

Operators should be able to see pending, leased, failed, and dead summary/extraction counts without querying SQLite directly.

## Suggested fix

- Add summary and extraction queue counts to `signet status`, `/api/status`, diagnostics, and readiness health.
- Degrade readiness when dead/backlogged work exceeds configurable thresholds.
- Include oldest job age, last error, and effective provider/block reason.
- Add supported repair commands with dry-run/preview for:
  - retrying selected dead jobs,
  - cancelling obsolete jobs,
  - pruning old terminal jobs while preserving provenance.

## Regression test

Seed dead and stale leased jobs. Assert status/doctor reports counts, oldest age, and degraded readiness; verify repair preview does not mutate data.

## Related
#908 — status conflates configured/resolved/effective/running states (root cause of hidden backlogs)
#902 — session TTL eviction drops tracker state without checkpoint (another hidden pipeline failure)

- #913 (unified dreaming) — summary-worker removal deletes this bug; its requirements are inherited by the unified ingest queue and §8 queue visibility.

### PR DESCRIPTION
**Pull request by aaf2tbz**
## Summary

Closes #901. A live `0.147.1` database contained 92 completed summary jobs and 1,667 dead summary jobs with no operator-visible signal. This PR surfaces per-queue health on daemon/CLI status paths and adds dry-run-first repair commands for requeue, cancel, and prune.

The branch is rebased onto current `main` (`v0.148.0`). TypeScript and Rust now share the same safety contract: bounded selection before mutation, transactional writes, migration-owned provenance tables, explicit admin gating, and replay coverage.

## Changes

- Adds `GET /api/diagnostics/queue` and `POST /api/diagnostics/queue/repair`.
- Adds `signet repair queue {requeue|cancel|prune}` with dry-run by default and explicit `--apply` mutation.
- Adds per-queue memory/summary/extraction counts and oldest-dead details to status surfaces.
- Adds migration 088 `job_cancellations` for audit-preserving soft cancellation.
- Adds migration 089 `job_archive` for provenance-preserving pruning.
- Mirrors both endpoints, migrations, response casing, and repair behavior in daemon-rs.
- Adds a 30-second per-database cache for expanded queue diagnostics used by status surfaces; explicit queue diagnostics bypass it and successful repairs invalidate it.
- Reuses the shared TypeScript `tableExists` helper instead of adding queue-local copies.

## Review fixes

Addressed NicholaiVogel's review:

1. Rust requeue selects bounded IDs first, then uses `UPDATE ... WHERE id IN (...)`; it no longer depends on SQLite `UPDATE ... LIMIT`.
2. Both Rust queue endpoints are included in the admin-required path set, with unit and live team-auth coverage.
3. Cancel/prune select `id, status` by name, so `summary_jobs.transcript` can never be recorded as `status_before`.
4. Rust no longer creates audit tables from the request path. The migration reconciler owns both tables and indexes, and the handler fails with a migration error if required schema is absent.
5. Explicit `olderThanMs` and `retentionMs` values are honored instead of being raised to the defaults.
6. Cancel/prune selection is globally bounded across both queue tables and runs in one transaction.
7. Rust provenance stores the complete source-row payload and preserves actor/reason/request headers.
8. Direct tests cover threshold scoring, requeue selection, mixed-schema status lookup, full payload capture, retention override, missing migration behavior, cancel, prune, and admin gating.
9. The POST repair route now has built-daemon contract replay coverage and is `native-rust-replay-proven`.
10. Rebase fallout from the already-landed Rust integrity/rebuild routes was removed; duplicate declarations no longer remain.

No new `summaryJobExists` implementation remains in this branch after the rebase. The pre-existing session recovery helper is not duplicated by this PR; cross-branch consolidation with #937 can use the shared helper when that branch is rebased without introducing another copy here.

## Packages affected

- [x] `@signet/core`
- [x] `@signet/daemon`
- [x] `@signet/cli`
- [x] `platform/daemon-rs`

## PR readiness

- [x] Spec alignment validated (`INDEX.md` + `dependencies.yaml`)
- [x] Agent scoping verified on all new/changed data queries
- [x] Input/config validation and bounds checks added
- [x] Error handling and fallback paths tested (no silent swallow)
- [x] Security checks applied to admin/mutation endpoints
- [x] Docs updated for API/spec/status changes
- [x] Regression tests added for each bug fix
- [x] Lint/typecheck/tests pass locally

## Migration notes

- [x] Migration is idempotent — migrations 088 and 089 use guarded DDL and own the audit/archive schema.
- [x] Rust's migration reconciler creates and stamps the same artifacts for Rust-first installations.
- [x] Apply paths fail closed when required audit schema is absent; diagnostics remain read-only and tolerate missing queue tables.
- [x] Daemon Rust parity reviewed or explicitly N/A.
- [x] Rollback / compatibility note included in PR description: older daemons ignore the additive tables and fields; rolling back removes the new repair/status surfaces but does not alter existing queue schemas. Operators should retain audit/archive tables if provenance history is needed.

## Validation

- [x] Focused TypeScript diagnostics/repair/CLI tests: 42 passed.
- [x] TypeScript migration + repair-action tests: 86 passed.
- [x] Biome clean on touched TypeScript files.
- [x] `@signet/daemon` production build passed.
- [x] Rust `signet-core` library: 77 passed.
- [x] Rust queue repair unit tests: 3 passed.
- [x] Rust admin-path unit test passed.
- [x] Rust daemon compile/check passed.
- [x] Built-daemon `diagnostics_endpoints` contract test passed, including bounded requeue, cancellation audit status, and full transcript payload provenance.
- [x] Built-daemon team-auth contract test passed, including operator denial and admin access for queue diagnostics/repair.
- [x] Route parity manifest is current: 333 routes; both queue routes are `native-rust-replay-proven`. The one `missing` route is `main`'s intentionally TypeScript-only `/api/inference/catalog`; `main` retired the hard-fail parity workflow in #962 rather than adding a misleading Rust stub.
- [x] Focused CLI queue tests passed. The repository-wide CLI suite still has unrelated environment/architecture and missing-workspace-package failures; the queue test file is green.
- [x] Repository TypeScript checking still reports pre-existing errors; no errors remain in the new diagnostics queue, repair route, or queue test code.

## AI disclosure

- [x] AI tools were used (see `Assisted-by` tags in commits).

### CODE REVIEW

File: platform/daemon/src/repair-actions.ts [hunk 1]
```diff
@@ -2285,3 +2353,864 @@ export async function rebuildDerivedIndexes(
 		summary: parts.join(" · "),
 	};
 }
+
+// ---------------------------------------------------------------------------
+// Internal helpers shared by selective repair actions (issue #901)
+// ---------------------------------------------------------------------------
+
+function tableExistsInternal(db: ReadDb, name: string): boolean {
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/repair-actions.ts]: Fixed in 092228f — the duplicate `tableExistsInternal` declaration was removed during the repair-actions helper dedupe; only one declaration remains (verified: no duplicate function names in `platform/daemon/src/repair-actions.ts`).

File: platform/daemon/src/repair-actions.ts [hunk 2]
```diff
@@ -2285,3 +2353,864 @@ export async function rebuildDerivedIndexes(
...
+	readonly totalMatching: number;
+}
+
+/**
+ * Build a parameterized SELECT that enumerates dead rows matching the
+ * filter. The `placeholderIdColumn` arg identifies the column that
+ * uniquely identifies the row inside the table — `memory_id` for
+ * `memory_jobs`, `null` for `summary_jobs` whose primary key is `id`.
+ */
+function buildDeadRequeueSql(
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/repair-actions.ts:2371]: Fixed in 092228f — duplicate `buildDeadRequeueSql` declaration removed; a single declaration remains and all call sites use it.

File: platform/daemon/src/repair-actions.ts [hunk 3]
```diff
@@ -2285,3 +2353,864 @@ export async function rebuildDerivedIndexes(
...
+	readonly rows: readonly CancelPruneMatchRow[];
+	readonly totalMatching: number;
+}
+
+/**
+ * Build a parameterized SELECT for `cancelObsoleteJobs` and
+ * `pruneTerminalJobs`. Captures the full row as JSON so the action can
+ * copy it to the audit/archive table inside the same write tx.
+ */
+function buildCancelPruneSql(
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/repair-actions.ts]: Fixed in 092228f — duplicate `buildCancelPruneSql` declaration removed during the repair-actions helper dedupe.

File: platform/daemon/src/repair-actions.ts [hunk 4]
```diff
@@ -2285,3 +2353,864 @@ export async function rebuildDerivedIndexes(
...
+	readonly totalMatching: number;
+}
+
+/**
+ * Cancel obsolete jobs by copying them to `job_cancellations` and flipping
+ * their source `status` to `cancelled`. Never hard-deletes the source
+ * row. Default selection: rows where `status IN ('dead','completed')`
+ * and `created_at < now - olderThanMs` (default 30 days).
+ */
+export function cancelObsoleteJobs(
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/repair-actions.ts]: Fixed in 092228f — duplicate `cancelObsoleteJobs` declaration removed; single exported declaration remains.

File: platform/daemon/src/repair-actions.ts [hunk 5]
```diff
@@ -2285,3 +2353,864 @@ export async function rebuildDerivedIndexes(
...
+	readonly preview: readonly string[];
+	readonly totalMatching: number;
+}
+
+/**
+ * Prune terminal jobs (cancelled / completed / dead) older than the
+ * configured retention window. Before delete, copies the full row to
+ * `job_archive` to preserve provenance. Hard cap is 1000 rows per call.
+ */
+export function pruneTerminalJobs(
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/repair-actions.ts]: Fixed in 092228f — duplicate `pruneTerminalJobs` declaration removed; single exported declaration remains.

File: platform/daemon/src/routes/queue-diagnostics.ts [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		/cooldown active|hourly budget exhausted|denied by policy gate|autonomous\.|agents cannot trigger repairs/i.test(
+			result.message,
+		)
+	) {
+		return 429;
+	}
+	return 500;
+}
+
+function dispatchRepair(
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/routes/queue-diagnostics.ts]: Fixed in 092228f — unused `dispatchRepair` helper removed; the POST handler dispatches through `dispatchRepairWith` only.

File: platform/daemon/src/routes/queue-diagnostics.ts [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+	pruneTerminalJobs,
+	requeueDeadJobs,
+} from "../repair-actions.js";
+import { authConfig, repairLimiter } from "./state.js";
+
+function resolveAgentsDir(): string {
+	return process.env.SIGNET_AGENTS_DIR ?? process.env.AGENTS_DIR ?? `${process.env.HOME ?? "/tmp"}/.agents`;
+}
+
+function resolveLimiter() {
```
[the remark this answers is not in the document]
Comment (aaf2tbz) [reply platform/daemon/src/routes/queue-diagnostics.ts]: Fixed in 1e5a915 — unused `resolveLimiter` helper deleted from `platform/daemon/src/routes/queue-diagnostics.ts`. The handler uses the shared `limiter` (deps override ?? `repairLimiter` ?? `createRateLimiter()`) directly. `bun test src/routes/queue-diagnostics` passes (7/7).

**Comment by NicholaiVogel** [comment]
Reviewed the full diff. The TS side is solid: dry-run-by-default, provenance-preserving cancel/prune, clean CLI plumbing. But the Rust side has three critical bugs and the same duplication pattern across multiple files. Requesting changes before merge.

## Critical bugs

### 1. Rust requeue apply uses UPDATE with LIMIT (runtime crash)

repair.rs line 4258-4265 builds:

UPDATE memory_jobs SET status='pending', attempts=0, updated_at=datetime('now') WHERE ... LIMIT {limit}

SQLite UPDATE with LIMIT requires the SQLITE_ENABLE_UPDATE_DELETE_LIMIT compile flag. The project uses rusqlite with the bundled feature and does not define this flag anywhere. This statement will fail at runtime with a parse error on every real requeue apply. The TS path avoids this by selecting IDs first, then doing UPDATE with a WHERE id IN (...) clause.

### 2. Rust diagnostic routes are not admin-gated (security hole)

The two new routes /api/diagnostics/queue and /api/diagnostics/queue/repair are not in the is_admin_required_path list (auth/middleware.rs line 224). The TS side enforces requirePermission('admin') on both routes. Under hybrid or local mode, any unauthenticated localhost caller can read full queue diagnostics and dispatch cancel, prune, or requeue operations.

### 3. Rust cancel/prune reads wrong column for summary_jobs status (silent data corruption)

repair.rs line 4375 does row.get::String(4) to read status_before from a SELECT * result. For memory_jobs, column index 4 is status. For summary_jobs, column index 4 is transcript and status is at index 5 (per migration 009). The audit rows for summary_jobs will contain transcript text in the status_before field. The TS side avoids this by reading row.status by name from a Record.

## Duplication and code smell

### ensure_audit_table bypasses migrations

repair.rs line 4295 calls ensure_audit_table(conn, 'job_cancellations') inside the apply path, which does CREATE TABLE IF NOT EXISTS. The TS side throws if the table is missing, which is correct. The Rust auto-create means the migration system is no longer the single source of truth for schema. If someone rolls back migrations 088 and 089 but the Rust daemon has already created the tables, the schema is in an inconsistent state.

### SELECT * with positional column indexing across two tables with different schemas

The cancel/prune path does SELECT * FROM {source_table} and then reads columns by index. memory_jobs and summary_jobs have different column orders. This is what caused bug 3. The correct approach is SELECT id, status FROM {source_table} with named columns, matching how the TS side reads by key.

### summaryJobExists duplicated three times

The same logic for checking whether a job exists (with session_id vs session_key fallback, agent_id column detection) appears in: (1) session-end-recovery.ts as summaryJobExistsForRecoveredSession, (2) the new session-ttl-finalization.ts from PR 937 as summaryJobExists, and (3) this PR as queue_table_exists and the inline checks in the requeue path. Three implementations of the same query.

### tableExists duplicated four times

The helper that does SELECT name FROM sqlite_master WHERE type='table' AND name=? is defined locally in: (1) this PR's repair.rs, (2) session-ttl-finalization.ts from PR 937, (3) diagnostics.ts, (4) session-end-recovery.ts. This should be a shared utility.

### retentionMs clamp ignores caller intent

repair.rs line 4340 clamps retention_ms with .max(retention_ms_default). A caller who explicitly passes retentionMs of 86400000 (1 day) for prune still gets the 90-day default. The TS side honors the caller value. Either the docs are wrong or the clamp is wrong.

### Performance regression on /api/status

getQueueHealth (diagnostics.ts) now issues 6 additional SELECTs per invocation: 3 getQueueCounts calls (each doing tableExists on sqlite_master plus column PRAGMAs plus the aggregate query) and 3 getOldestDeadJob calls. These run on every status and diagnostics call with no caching, despite the existing STATUS_CACHE_TTL mechanism in routes/utils.ts.

## Requested changes

1. Fix the three critical bugs (UPDATE...LIMIT, admin gate, column index).
2. Use SELECT id, status with named columns instead of SELECT * with positional indexing.
3. Remove ensure_audit_table and let the migration system own schema creation.
4. Extract tableExists and summaryJobExists into shared utilities.
5. Add caching or a query flag for the per-queue expansion in getQueueHealth.
6. Add unit tests for scoreCountsWithThresholds, cancelObsoleteJobs, and pruneTerminalJobs (currently zero direct tests).
7. Add the missing auth-gate test that was promised in the plan.

**Comment by aaf2tbz** [comment]
Addressed the full review in rebased follow-up commit `bbaec976`.

The three critical Rust bugs are fixed:
- requeue now selects bounded IDs and updates with `WHERE id IN (...)` instead of `UPDATE ... LIMIT`
- both queue routes are admin-gated in Rust, with unit coverage and a live team-mode replay proving operator `403` / admin `200`
- cancel/prune select `id, status` by name; the replay verifies a summary job records `status_before = dead`, not its transcript

The requested structural fixes are also in:
- removed request-path DDL; Rust's migration reconciler owns and stamps 088/089 and missing audit schema fails closed
- removed the queue-local TypeScript `tableExists` copies in favor of the shared helper, and reused the existing Rust table helper
- explicit retention overrides are honored
- expanded queue health is cached for status callers, fresh diagnostics bypass the cache, and successful repairs invalidate it
- direct tests now cover threshold scoring, cancel/prune, bounded requeue, mixed table schemas, full payload provenance, migration ownership, and auth gates

While resolving the branch conflicts, I also removed the duplicated Rust `integrity_check` / `rebuild_indexes` declarations left by the now-landed parity commit, regenerated route parity, and rebased onto `v0.147.21`. Both queue routes are now `native-rust-replay-proven`.

On `summaryJobExists`: after rebasing, this PR introduces no such implementation. The existing recovery helper remains pre-existing; adding another wrapper here would create the duplication the review warned against. #937 can consume/extract the shared helper when it rebases.

Validation details are updated in the PR description. The built-daemon repair replay and built-daemon team-auth replay both pass.

**Review by NicholaiVogel — APPROVED** [review]
Approving after full review (changed functions + callers + the transaction/dedup contracts traced, not diff-only), a migration renumber to resolve a collision with the already-landed #998, and a clean rebase onto 0.151.0 (17/17 CI green).

What this adds (#901): surfaces per-queue dead-job health (GET /api/diagnostics/queue, status surfaces, signet status) and adds dry-run-first repair actions (POST /api/diagnostics/queue/repair, `signet repair queue {requeue|cancel|prune}`). Closes the gap where 1,667 dead summary jobs had zero operator-visible signal.

Migration renumber (the one real change I made on top of Avery's work):
- This branch originally declared migrations 088-job-cancellations + 089-job-archive, but #998 landed first and took 088-transcript-recovery-files. Colliding version numbers would be rejected by the migrator.
- Renumbered to 089-job-cancellations + 090-job-archive across TS (file names, imports, version fields, header comments), Rust (TS_JOB_CANCELLATIONS_VERSION/_ARCHIVE_VERSION constants + the parity test assertions), and docs (PIPELINE.md, health-status.md, plans doc). 088-transcript-recovery-files is preserved from main.
- Proven correct: migration framework test (44 pass, includes duplicate-version rejection + all-tables-exist), Rust parity tests (records_ts_parity_migrations_with_artifact_backing, rust_created_schema_satisfies_typescript_migration_artifacts — both assert 89/90 stamped), cargo check clean, no stale 088/089-job refs anywhere, TS registry is contiguous 1–90 with no duplicates.

Mutation safety (the highest-risk logic) — verified sound:
- All three actions select bounded IDs first (LIMIT maxBatch, hard cap 1000), then UPDATE … WHERE id IN (…) inside a single withWriteTx (BEGIN IMMEDIATE). No dependence on SQLite UPDATE … LIMIT.
- Provenance preserved: cancel copies full row → job_cancellations + flips status to 'cancelled' (soft); prune copies full row → job_archive before delete.
- Fails closed: cancel/prune throw "table missing; run migrations" if schema absent — tables are never created from the request path (migration reconciler owns them, TS + Rust).

Security: both endpoints admin-gated (TS requirePermission("admin") + Rust is_admin_required_path, with unit test); dry-run by default (dryRun !== false); rate-limited + policy-gated; all inputs parameterized (no injection).

One fix from autoreview (merged here with a regression test): the POST handler previously wrapped only body parsing in try/catch, so a thrown repair error (missing-migration / closed DbAccessor / SQLite error) escaped as an unstructured 500. Now mirrors the GET sibling and Rust parity handler — returns the documented {action, success:false, affected:0, message} RepairResult. Regression test added ("returns a structured result when a repair action throws").

Non-blocking notes (left as-is, design-consistent):
- ids array is unbounded in the IN-list builder — admin-gated robustness nit, not exploitable.
- actorType comes from a client header (x-signet-actor-type) — consistent with the local-first/hybrid-localhost trust model and admin-gated.
- prune intentionally ignores olderThanMs in favor of retentionMs; TS and Rust agree, CLI maps --older-than to retentionMs for prune. Deliberate contract.

Proof: 35/35 TS pass (queue-diagnostics incl. new regression + diagnostics), 8/8 CLI repair-queue, Rust compiles + 3 parity tests pass. Rebased onto 0.151.0; functional files byte-identical to the reviewed branch.
